hooks/useRailgunTx.tsx

```diff
@@ -0,0 +1,188 @@
+      maxPriorityFeePerGasString: maxPriorityFeePerGas!.toHexString(), // Current gas Max Priority Fee
+    };
+
+    const { serializedTransaction, error } = await populateShieldBaseToken(
+      network,
+      recipient,
+      shieldPrivateKey,
+      wrappedERC20Amount,
+      gasDetailsSerialized
+    );
+    if (error) {
+      throw error;
+    }
+
+    const { chain } = NETWORK_CONFIG[network];
+
+    const transactionRequest: ethers.providers.TransactionRequest =
+      deserializeTransaction(
+        serializedTransaction!,
+        undefined, // nonce (optional)
+        chain.id
+      );
+
+    // Public wallet to shield from.
+    transactionRequest.from = address;
+
+    return signer?.sendTransaction(transactionRequest);
+  };
+
+  const shieldToken = async ({
+    tokenAddress,
+    tokenAmount,
+    tokenDecimals,
+    recipient,
+  }: {
+    tokenAddress: string;
+    tokenAmount: string;
+    tokenDecimals: number;
+    recipient: string;
+  }) => {
+    // Public wallet to shield from.
+    const fromWalletAddress = address as `0x{string}`;
+
+    // The shieldPrivateKey enables the sender to decrypt
+    // the receiver's address in the future.
+    const shieldPrivateKey = await getShieldPrivateKey();
+
+    // Formatted token amounts and recipients.
+    const erc20AmountRecipients: RailgunERC20AmountRecipient[] = [
+      {
+        tokenAddress: tokenAddress!,
+        amountString: ethers.utils
+          .parseUnits(tokenAmount!, tokenDecimals)
+          .toHexString(), // must be hex
+        recipientAddress: recipient!, // RAILGUN address
+      },
+    ];
+
+    const { gasEstimateString, error: err } = await gasEstimateForShield(
+      network,
+      shieldPrivateKey,
+      erc20AmountRecipients,
+      [], // nftAmountRecipients
+      fromWalletAddress
+    );
+
+    if (err) {
+      throw err;
+    }
+
+    const { maxFeePerGas, maxPriorityFeePerGas } = await provider.getFeeData();
+    const gasDetailsSerialized: TransactionGasDetailsSerialized = {
+      evmGasType: EVMGasType.Type2, // Depends on the chain (BNB uses type 0)
+      gasEstimateString: gasEstimateString!, // Output from gasEstimateForShield
+      maxFeePerGasString: maxFeePerGas!.toHexString(), // Current gas Max Fee
+      maxPriorityFeePerGasString: maxPriorityFeePerGas!.toHexString(), // Current gas Max Priority Fee
+    };
+
+    const { serializedTransaction, error } = await populateShield(
+      network,
+      shieldPrivateKey,
+      erc20AmountRecipients,
+      [], // nftAmountRecipients
+      gasDetailsSerialized
+    );
+    if (error) {
+      throw error;
+    }
+
+    const { chain } = NETWORK_CONFIG[network];
+
+    const transactionRequest: ethers.providers.TransactionRequest =
+      deserializeTransaction(
+        serializedTransaction as string,
+        undefined, // nonce (optional)
+        chain.id
+      );
+
+    // Public wallet to shield from.
+    transactionRequest.from = address;
```

Comment:
Feels like a lot of duplicate code between the native token and ERC20 methods. Wonder if we can/should dedupe a bit?

Reply:
I looked into deduping, it got a bit convoluted. For simplicity i thought we might PR this version, and if we see the opportunity to refactor then we could go for it. What do you think?